strawberryfields/engine.py

```diff
@@ -341,11 +378,13 @@ def _init_backend(self, init_num_subsystems):
 
     def _run_program(self, prog, **kwargs):
         applied = []
+        values = []
         for cmd in prog.circuit:
             try:
-                # try to apply it to the backend
-                # NOTE we could also handle storing measured vals here
-                cmd.op.apply(cmd.reg, self.backend, **kwargs)
+                # try to apply it to the backend and, if op is a measurement, store it in values
+                val = cmd.op.apply(cmd.reg, self.backend, **kwargs)
```

Comment:
Question related to my comment above re: `RuntimeError("Modes can only be measured once inside a circuit.")`.

Previously, if we ran the program 
```python
with prog.context as q:
    Sgate(0.1) | q[0]
    MeasureX | q[0]  # resets q[0] to the vacuum
    Xgate(q[0].par) | q[0]
    MeasureFock() | q[0]
```

we extracted the returned samples by doing `[r.val for r in regrefs]`. So even though `q[0]` would be measured twice, only the final measured value would be provided (as the measured value of each regref gets overwritten with each measurement).

Now, however, the measurement values are returned directly from the backends, so `values` will have _two_ measured values of `q[0]` in the list, triggering the above exception.

We have two choices here:

1. Keep the existing behaviour, by having `combine_samples` simply take the _last_ value corresponding to modes that appear twice.

2. Allow modes to be measured twice within `combine_samples`, e.g., the above program would give

   ```python
   >>> results.samples
   array([[0.7, 1], [0.1, 0], ...])
   ```
   where column 0 corresponds to the first temporal q[0] measurement, and column 1 corresponding to the second temporal q[1] measurement.

   This changes are `(shots, modes)` convention into `(shots, measurements)`.

---

Side note: I'm realizing now that there will be a second consequence of this PR; samples will now be returned corresponding to the order that the user performed the measurements. E.g., if the user does

```python
MeasureFock | q[1]
MeasureFock | q[0]
```

then the returned samples will be ordered [mode 1, mode 0].

I think this makes sense to some degree, _but_ the more I think about this (and in addition with PR #399), the more I think it makes sense to turn `results.samples` into a class object, so we can provide the additional metainformation.

```pycon
>>> samples.num_modes
4
>>> samples.num_shots
500
>>> samples.num_measurements
6
>>> samples.modes
[0, 2, 1, 3, 1, 0]
>>> samples.mtypes
['Homodyne', 'Fock', ...]
>>> samples.values.shape
[500, 6]
```

Reply:
> 1. Keep the existing behaviour, by having `combine_samples` simply take the _last_ value corresponding to modes that appear twice.

This would make more sense with how things work in this PR.

> 2. Allow modes to be measured twice within `combine_samples`

I'm not sure if this is actually something that's wanted/needed. Perhaps it's nice to have, although it would completely change what is returned to the user.

> Side note: I'm realizing now that there will be a second consequence of this PR; samples will now be returned corresponding to the order that the user performed the measurements. 

This shouldn't be the case since the mode order is taken into account when combining the samples; e.g. `samples = [array([[0, 1]]), array([[2]])]` and `mode_order = [array([0, 3]), array([1])]` would return `array([[0, 2, 1]])` to the user, which would be in ascending mode order (as the docstring for `combine_samples` also states).

> I think this makes sense to some degree, _but_ the more I think about this (and in addition with PR #399), the more I think it makes sense to turn `results.samples` into a class object, so we can provide the additional metainformation.

I think I agree with you, but this maybe should be another PR, right? In that case we could leave this issue to be handled in that PR instead (including `RuntimeError("Modes can only be measured once inside a circuit.")`. Or what does everyone think?

Reply:
> > 1. Allow modes to be measured twice within `combine_samples`
> 
> I'm not sure if this is actually something that's wanted/needed. Perhaps it's nice to have, although it would completely change what is returned to the user.
> 

Somewhat related, but I wanted to reiterate that measuring a mode multiple times is a supported feature of SF. It's not a new feature that's being considered. We have to consider the present PR in that context. If we feel the need to change a feature that's been in the library from the start, probably best to do as a separate discussion/PR

Reply:
I agree that this might be a bit out-of-scope for this PR. The change mostly happened by mistake to be honest, since the changes caused other issues when combining the samples when measuring the same mode twice (thus this was the easy fix). Should've checked how it worked before rather than just removing that possibility. :slightly_frowning_face: